dev/AppLifecycle/AppInstance.cpp

```diff
@@ -0,0 +1,415 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include <pch.h>
+#include <AppInstance.h>
+#include <Microsoft.Windows.AppLifecycle.AppInstance.g.cpp>
+
+#include "ActivationRegistrationManager.h"
+#include "LaunchActivatedEventArgs.h"
+#include "ProtocolActivatedEventArgs.h"
+#include "FileActivatedEventArgs.h"
+#include "Association.h"
+#include "ExtensionContract.h"
+
+using namespace winrt;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::ApplicationModel::Activation;
+
+namespace winrt::Microsoft::Windows::AppLifecycle::implementation
+{
+    INIT_ONCE AppInstance::s_initOnce{};
+    winrt::com_ptr<AppInstance> AppInstance::s_current;
+
+    std::tuple<std::wstring, std::wstring> ParseCommandLine(std::wstring commandLine)
+    {
+        auto argsStart = commandLine.rfind(c_argumentPrefix);
+        if (argsStart == std::wstring::npos)
+        {
+            return { L"", L"" };
+        }
+
+        argsStart += 4;
+
+        // We explicitly use find_first_of here, so that the resulting data may contain : as a valid character.
+        auto argsEnd = commandLine.find_first_of(L":", argsStart);
+        if (argsEnd == std::wstring::npos)
+        {
+            return { L"", L"" };
+        }
+
+        if (argsStart > argsEnd)
+        {
+            throw std::overflow_error("commandLine");
+        }
+
+        auto argsLength = argsEnd - argsStart;
+        auto dataStart = argsEnd + 1;
+
+        return { commandLine.substr(argsStart, argsLength), commandLine.substr(dataStart) };
+    }
+
+    std::tuple<ExtendedActivationKind, winrt::Windows::Foundation::IInspectable> GetEncodedLaunchActivatedEventArgs(IProtocolActivatedEventArgs const& args)
+    {
+        auto kind = ExtendedActivationKind::Protocol;
+        winrt::Windows::Foundation::IInspectable data;
+
+        std::tie(kind, data) = DecodeActivatedEventArgs(args.Uri());
+
+        // Let the caller args pass through if nothing was determined here.
+        if (data == nullptr)
+        {
+            data = args;
+        }
+        
+        return { kind, data };
+    }
+
+    AppInstance::AppInstance(uint32_t processId)
+    {
+        m_processId = processId;
+        m_isCurrent = (GetCurrentProcessId() == processId);
+
+        m_moduleName = ComputeAppId();
+        m_processName = wil::str_printf<std::wstring>(L"%s_%d", m_moduleName.c_str(), processId);
```

Comment:
Is this scheme potentially compromised by pid recycling?  (unlikely but wondering if you considered this)